File: qc-application/src/main/java/com/eager/questioncloud/application/api/payment/question/implement/QuestionPaymentProcessor.java
package com.eager.questioncloud.application.api.payment.question.implement;

import com.eager.questioncloud.core.domain.payment.infrastructure.QuestionOrderRepository;
import com.eager.questioncloud.core.domain.payment.infrastructure.QuestionPaymentRepository;
import com.eager.questioncloud.core.domain.payment.model.QuestionPayment;
import com.eager.questioncloud.core.domain.point.implement.UserPointManager;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Component;
import org.springframework.transaction.annotation.Transactional;

@Component
@RequiredArgsConstructor
public class QuestionPaymentProcessor {
    private final QuestionPaymentRepository questionPaymentRepository;
    private final QuestionOrderRepository questionOrderRepository;
    private final QuestionPaymentCouponProcessor questionPaymentCouponProcessor;
    private final UserPointManager userPointManager;

    @Transactional
    public QuestionPayment processQuestionPayment(QuestionPayment questionPayment) {
        questionPaymentCouponProcessor.applyCoupon(questionPayment);
        userPointManager.usePoint(questionPayment.getUserId(), questionPayment.getAmount());
        questionOrderRepository.save(questionPayment.getOrder());
        return questionPaymentRepository.save(questionPayment);
    }
}

File: qc-application/src/main/kotlin/com/eager/questioncloud/application/api/payment/question/implement/QuestionPaymentProcessor.kt
package com.eager.questioncloud.application.api.payment.question.implement

import com.eager.questioncloud.core.domain.payment.infrastructure.repository.QuestionPaymentRepository
import com.eager.questioncloud.core.domain.payment.model.QuestionPayment
import com.eager.questioncloud.core.domain.point.implement.UserPointManager
import org.springframework.stereotype.Component
import org.springframework.transaction.annotation.Transactional

@Component
class QuestionPaymentProcessor(
    private val questionPaymentRepository: QuestionPaymentRepository,
    private val questionPaymentCouponProcessor: QuestionPaymentCouponProcessor,
    private val userPointManager: UserPointManager
) {
    @Transactional
    fun payment(questionPayment: QuestionPayment) {
        questionPaymentCouponProcessor.applyCoupon(questionPayment)
        userPointManager.usePoint(questionPayment.userId, questionPayment.amount)
        questionPaymentRepository.save(questionPayment)
    }
}
